code/HasCanonicalEditLink.php

```diff
@@ -0,0 +1,152 @@
+<?php
+
+namespace SilverStripe\Admin;
+
+use LogicException;
+use SilverStripe\CMS\Controllers\CMSMain;
+use SilverStripe\Control\Controller;
+use SilverStripe\Control\Director;
+use SilverStripe\Forms\FieldList;
+use SilverStripe\Forms\GridField\GridField;
+use SilverStripe\Forms\GridField\GridFieldDetailForm;
+use SilverStripe\ORM\DataObject;
+
+/**
+ * A trait that automatically generates a CMS edit link for DataObjects even if
+ * they are canonically edited in some nested {@link GridField}.
+ * Designed to be used in conjunction with the {@link CMSPreviewable} interface.
+ *
+ * For nested relations (e.g. a DataObject managed in a GridField of another DataObject)
+ * you can apply this trait to both the parent and the child object and the links
+ * will chain down the nested `GridField`s to the root canonical edit owner.
+ *
+ * You must set a canonical_edit_owner config variable which defines the canonical
+ * owner for this class.
+ * e.g. set this to a {@link LeftAndMain} class:
+ * private static string canonical_edit_owner = MyModelAdmin::class;
+ * or to a has_one relation:
+ * private static string canonical_edit_owner = 'Parent';
+ *
+ * Note that the canonical edit owner must implement a getEditLinkForObject() method.
+ *
+ * If the canonical edit owner is a has_one relation, the class on the other end
+ * of the relation must have a CMSEditLink() method.
+ */
+trait HasCanonicalEditLink
+{
+    /**
+     * Get the admin or DataObject which owns this object for CMS editing purposes.
+     *
+     * @return LeftAndMain|DataObject|null
+     */
+    public function getCanonicalEditOwner()
+    {
+        $ownerType = static::config()->get('canonical_edit_owner');
+        if (is_subclass_of($ownerType, LeftAndMain::class)) {
+            return $ownerType::singleton();
+        }
+        return $this->getComponent($ownerType);
+    }
+
+    /**
+     * Get the link for editing an object from the CMS edit form of this object.
+     * @throws LogicException if a link cannot be established
+     * e.g. if the object is not in a has_many relation or not edited inside a GridField.
+     */
+    public function getEditLinkForObject(DataObject $obj, string $reciprocalRelation): string
+    {
+        $fields = $this->getCMSFields();
+        $link = $this->getLinkForRelation($this->hasMany(false), $obj, $reciprocalRelation, $fields);
+        if (!$link) {
+            throw new LogicException('Could not produce an edit link for the passed object.');
+        }
+        return $link;
+    }
+
+    /**
+     * Get a link to edit this page in the CMS.
+     */
+    public function CMSEditLink(): string
```

Comment:
`File` already has `CMSEditLink` method, would this be a problem when applied to e.g. Image or other class extending `File`?

Reply:
I can't think of literally any reason someone would do that - but if they did, I'm pretty sure the trait implementation would take precedence.